### PR TITLE
[6346] Allow skipping changelog test from description

### PR DESCRIPTION
**Pull request by srbarrios**
## What does this PR do?

Related to Task: https://github.com/SUSE/spacewalk/issues/6346

It changes the way we re-run tests and we specify if changelog must be checked. we will use the checkboxes you see bellow: 

- [ ] Re-run test "changelog_test"

- [ ] No changelog needed

In the first case, gitarro will change from checked to unchecked after re-run th test.

### CODE REVIEW

File: tools/scripts/changelog_test.rb [hunk 1]
```diff
@@ -30,19 +30,29 @@ def initialize(client, repo, pr_num)
...
   def magic_comment?
     @client.issue_comments(repo, pr_num).any? do |com|
       com.body.include?('no changelog needed')
     end
   end
 
+  def no_changelog_needed?
+    return false if pr_num.nil?
+    pr = @client.pull_request(repo, pr_num)
+    return true unless pr.body.include? "[x] No changelog needed"
```
[the remark this answers is not in the document]
Comment (srbarrios) [reply tools/scripts/changelog_test.rb]: nope, is not, and yeah we can make it case insensitive, good point :+1:

**Comment by srbarrios** [comment]
Cool! Thank you @MalloZup !

File: lib/gitarro/backend.rb [hunk 2]
```diff
@@ -244,19 +244,32 @@ def reviewed_pr?(comm_st, pr)
...
     false
   end
 
+  # this function will check if the PR contains a checkbox
+  # for retrigger all the tests.
+  def retriggered_by_checkbox?(pr, context)
+    return false if pr.nil?
+    return false unless pr.body.include? "[x] Re-run test \"#{context}\""
+    puts "Re-run test \"#{context}\""
+    new_pr_body = pr.body.gsub("[x] Re-run test \"#{context}\"", "[ ] Re-run test \"#{context}\"")
```
[the remark this answers is not in the document]
Comment (srbarrios) [reply lib/gitarro/backend.rb]: I tried to move to insensitive, doing .downcase of both operands, but it's a bit harderwhen we want also replace it later to uncheck the checkbox.. The idea is to modify our template when we create a PR, to already have these checkboxes available, we shouldn't have much problems with case sensitive... anyway gonna think a bit more about a solution ;)

File: doc/ADVANCED.md [hunk 3]
```diff
@@ -31,6 +31,9 @@ In order to retrigger a specific test, you need to add a comment to the PR.
 **PLESE NOTE**: gitarro will delete the comment where you write the retriggering command,
 so make sure that you use a separate comment for this.
 
+**NEW!**: You can also create a checkbox in the PR description with the text : Re-run test "<test_name>",
```
Comment (MalloZup) [inline doc/ADVANCED.md]: `NEW` here seems not good. I mean in 5 months it will become old. 
More then this it add maintenance on documentation.

What is needed here is a `changelog.md`  with versioning. 
Since we have already the rubygem Gemfile, we can add this, so an user can grep for the `new` changes.

https://github.com/openSUSE/gitarro/issues/156

File: lib/gitarro/backend.rb [hunk 4]
```diff
@@ -244,19 +244,32 @@ def reviewed_pr?(comm_st, pr)
 
   # this function will check if the PR contains in comment the magic word
   # # for retrigger all the tests.
+  # TODO: To be deprecated when retriggered_by_checkbox? is welcomed by users
```
Comment (MalloZup) [inline lib/gitarro/backend.rb]: here 2 remarks:

1) TODO in code are will get lost and forgotten :smile: . So we don't want this here, but in a GitHub issue is better.

2) and more important: the usage of `gitarro` by SUSE-Manager doesn't dictate the direction of `gitarro` itself, which is a generic opensource project.
In this particolar case, if in the SUSE-Manager/Uyuni project you don't use the comment, this is a particular use-case, no need to deprecate the comment feature.

In any case the `comment` feature , is needed for retrigger failed test, so we cannot deprecate it, since they are 2 diff features.

File: lib/gitarro/backend.rb [hunk 5]
```diff
@@ -244,19 +244,32 @@ def reviewed_pr?(comm_st, pr)
...
         return true
       end
     end
     false
   end
 
+  # this function will check if the PR contains a checkbox
+  # for retrigger all the tests.
+  def retriggered_by_checkbox?(pr, context)
+    return false if pr.nil?
```
Comment (MalloZup) [inline lib/gitarro/backend.rb]: i don't think this is needed, a pr object isn't nil otherwise we catch it before.

Why did you add this check here?

File: lib/gitarro/backend.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (MalloZup) [reply lib/gitarro/backend.rb]: @mateiw  yop thx for this.

I'm openinig an issue for the case-insistive in general. i will fix once free time https://github.com/openSUSE/gitarro/issues/157

the solution is pretty simple:

string.downcase!

File: tools/scripts/changelog_test.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (MalloZup) [reply tools/scripts/changelog_test.rb]: https://github.com/openSUSE/gitarro/issues/157

File: tools/scripts/changelog_test.rb [hunk 6]
```diff
@@ -30,19 +30,29 @@ def initialize(client, repo, pr_num)
...
     # if the pr contains changes on .changes file, test ok
-    return true if pr_contains_changelog? || magic_comment?
+    return true if pr_contains_changelog? || no_changelog_needed? || magic_comment?
 
     false
   end
 
   private
 
+  #TODO: To be deprecated when no_changelog_needed? is welcomed by users
```
Comment (MalloZup) [inline tools/scripts/changelog_test.rb]: remove todo

File: doc/ADVANCED.md (hunk 3 above)
Comment (srbarrios) [reply doc/ADVANCED.md]: awesome, agree

File: lib/gitarro/backend.rb (hunk 4 above)
Comment (srbarrios) [reply lib/gitarro/backend.rb]: > "In any case the comment feature , is needed for retrigger failed test"

No need anymore with my PR, that's the idea ;)

File: doc/ADVANCED.md [hunk 7]
```diff
@@ -31,6 +31,9 @@ In order to retrigger a specific test, you need to add a comment to the PR.
 **PLESE NOTE**: gitarro will delete the comment where you write the retriggering command,
 so make sure that you use a separate comment for this.
 
+**NEW!**: You can also create a checkbox in the PR description with the text : Re-run test "<test_name>", 
+if it's checked gitarro will re-run the test and uncheck it.
```
Comment (MalloZup) [inline doc/ADVANCED.md]: this documentation is to much simple. 

I would add a concrete example of the feature, so it's more obvious. :sunflower:

File: lib/gitarro/backend.rb (hunk 4 above)
Comment (MalloZup) [reply lib/gitarro/backend.rb]: I don't think the checkbox should remove the comment. The comment workflow is more universal. kuberentes and other communities use this workflow.

File: lib/gitarro/backend.rb (hunk 5 above)
Comment (srbarrios) [reply lib/gitarro/backend.rb]: Totally right.

File: lib/gitarro/backend.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (srbarrios) [reply lib/gitarro/backend.rb]: Nope, is not. :dagger: 
See, the gsub

**Comment by MalloZup** [comment]
@srbarrios other then adressing the comments, take a look at travis and make rubocop happy.

https://travis-ci.org/openSUSE/gitarro/builds/474335708?utm_source=github_status&utm_medium=notification
:sunflower: 

Some things are unrelated to your PR but some not

**Comment by MalloZup** [comment]
@srbarrios  the checkbox idea in highlevel perspective is a clever one idea. ( allthough the comment trigger should stay)

For the `acceptance test` needed by this feature, take a look at this doc: 

https://github.com/openSUSE/gitarro/blob/master/doc/TESTS.md#acceptance-tests

File: lib/gitarro/backend.rb (hunk 4 above)
Comment (srbarrios) [reply lib/gitarro/backend.rb]: Nono I don't think we should remove it. Let's keep both, yes!
What I mean is that in Spacewalk, we will use the checkbox for re-trigger failure tests

File: doc/ADVANCED.md (hunk 7 above)
Comment (srbarrios) [reply doc/ADVANCED.md]: Yeah.. I let it that for the end, doc must be improved and probably a bit refactored at that section, agree.
Most important part for me is that we agree in code :)

File: lib/gitarro/backend.rb (hunk 4 above)
Comment (MalloZup) [reply lib/gitarro/backend.rb]: ok thx for clarification :) imho in spacewalk an user could choose to use both.
:sunflower:

**Comment by srbarrios** [comment]
> @srbarrios the checkbox idea in highlevel perspective is a clever one idea. ( allthough the comment trigger should stay)
> 
> For the `acceptance test` needed by this feature, take a look at this doc:
> 
> https://github.com/openSUSE/gitarro/blob/master/doc/TESTS.md#acceptance-tests

It should be fine now, all the tests are passing, rubocop is green and I changed the doc.
Take a look please

**Comment by MalloZup** [comment]
thx @srbarrios i will have look tmw

**Comment by MalloZup** [comment]
@srbarrios  please rebase the commits

**Comment by MalloZup** [comment]
@juliogonzalez   i let you release the `rubygem` and update the stuff on manager side

**Comment by srbarrios** [comment]
> holding to red , i will make it green once rebased.

Yep, Im on it, I had a conflict with CHANGELOG.md after rebase and is why you see more commits, trying to see how to fix it.

**Comment by srbarrios** [comment]
Ok, I can't, I think I'm forced to have two commits :/
As CHANGELOG was a new file added on master.

**Comment by MalloZup** [comment]
ok for 2 commits no pb

**Comment by srbarrios** [comment]
> ok for 2 commits no pb but you should rebase to `upstream/master` for avoding the merge commit

Yep, is what I did. But after rebase `upstream/master`, I though I would be able to rebase all in one commit, but doing it it show me the conflict with CHANGELOG.md, so I guess I will need to keep it in a separate commit

**Comment by MalloZup** [comment]
> > ok for 2 commits no pb but you should rebase to `upstream/master` for avoding the merge commit
> 
> Yep, is what I did. But after rebase `upstream/master`, I though I would be able to rebase all in one commit, but doing it it show me the conflict with CHANGELOG.md, so I guess I will need to keep it in a separate commit

try harder, resolve the conflict again and you will fix it. This is a common annoying issue on git but you can still fix it.
